### PR TITLE
Improve on tool input handling and add tests

### PR DESCRIPTION
**Pull request by olaservo**
## Motivation and Context

The ToolTab JSON input handling still has some quirks and testing it manually isn't very sustainable.  This PR adds the Jest testing framework to the client side of the MCP Inspector, enabling unit testing for the more complex client-side utilities and components. 

The intent here is to improve maintainability by:
1. Extracting utility functions from components into dedicated utility files
	- `jsonPathUtils.ts`: Handles JSON path operations (updating and retrieving values)
    - `schemaUtils.ts`: Handles schema-related operations (generating default values, formatting labels)
2. Adding unit tests for these utilities
3. Improving the DynamicJsonForm component with better error handling and state management, for better handling of edge cases (empty objects, missing schema properties).
4. Enhancing the JsonEditor component with better state management and formatting capabilities
5. Address this issue: https://github.com/modelcontextprotocol/inspector/issues/154

Another option to consider is using a JSON editing library which handles more of this stuff out of the box.  I prefer to avoid adding libraries which might be overkill and introduce their own overhead.  But it's still an option, if trying to cover all the more complex tool input testing scenarios doesn't seem sustainable in the long term.

## How Has This Been Tested?

The PR includes new Jest tests for the utility functions that were extracted from components:
- Tests for jsonPathUtils (updateValueAtPath, getValueAtPath)
- Tests for schemaUtils (generateDefaultValue, formatFieldLabel, validateValueAgainstSchema)

I also updated the GitHub workflow to run these tests automatically during CI.

After these changes I also confirmed that I could update JSON fields for object tool inputs, which appears to be broken right now.

## Breaking Changes

None.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist

- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io/)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling

### CODE REVIEW

**Comment by olaservo** [comment]
Hi @jspahrsummers @jerome3o-anthropic @ashwin-ant hoping to get a review on this one - sorry it got big, a few related issues + tests are bundled together here.  Thanks!

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -25,6 +25,12 @@ jobs:
       # Working around https://github.com/npm/cli/issues/4828
       # - run: npm ci
       - run: npm install --no-package-lock
+
+      - name: Run client tests
+        run: |
+          cd client
+          npm test
```
Comment (jspahrsummers) [inline .github/workflows/main.yml]: TIOLI:
```suggestion
        working-directory: ./client
        run: npm test
```

File: client/src/utils/__mocks__/DynamicJsonForm.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Mock for DynamicJsonForm that only exports the types needed for tests
```
Comment (jspahrsummers) [inline client/src/utils/__mocks__/DynamicJsonForm.ts]: What happens when importing the actual component in tests? Just wondering if we _need_ this file or if it's a structural choice.

File: client/src/components/ToolsTab.tsx [hunk 3]
```diff
@@ -214,7 +214,10 @@ const ToolsTab = ({
                               description: prop.description,
                               items: prop.items,
                             }}
-                            value={(params[key] as JsonValue) ?? {}}
+                            value={
+                              (params[key] as JsonValue) ??
+                              (prop.type === "array" ? [] : {})
```
Comment (jspahrsummers) [inline client/src/components/ToolsTab.tsx]: Should this use `generateDefaultValue`?

File: client/src/components/DynamicJsonForm.tsx [hunk 4]
```diff
@@ -43,29 +46,63 @@ const DynamicJsonForm = ({
...
-      case "object": {
-        const obj: JsonObject = {};
-        if (propSchema.properties) {
-          Object.entries(propSchema.properties).forEach(([key, prop]) => {
-            obj[key] = generateDefaultValue(prop);
-          });
+  // Create a ref to store the timeout ID
+  const timeoutRef = useRef<ReturnType<typeof setTimeout>>();
+
+  // Create a debounced function to update parent state
```
Comment (jspahrsummers) [inline client/src/components/DynamicJsonForm.tsx]: Just curious, why?

File: client/src/components/DynamicJsonForm.tsx [hunk 5]
```diff
@@ -103,21 +140,61 @@ const DynamicJsonForm = ({
...
     switch (propSchema.type) {
       case "string":
+        return (
+          <Input
+            type="text"
+            value={(currentValue as string) ?? ""}
+            onChange={(e) => {
+              const val = e.target.value;
+              if (!val && !propSchema.required) {
+                handleFieldChange(path, undefined);
```
Comment (jspahrsummers) [inline client/src/components/DynamicJsonForm.tsx:154]: Why don't we want to use `generateDefaultValue` for all of these?

**Review by jspahrsummers — DISMISSED** [review]
coolkid

Just a few questions, but nothing blocking—feel free to merge at will!

File: client/src/utils/__mocks__/DynamicJsonForm.ts (hunk 2 above)
Comment (olaservo) [reply client/src/utils/__mocks__/DynamicJsonForm.ts]: It wasn't actually needed anymore, I had temporarily created it while sorting out some test details but tests should run fine without it.

File: client/src/components/ToolsTab.tsx (hunk 3 above)
Comment (olaservo) [reply client/src/components/ToolsTab.tsx]: I went back and forth on this and ended up using `generateDefaultValue` here, with the intention of using the tool schema defaults rather than forcing any specific defaults here.

File: client/src/components/DynamicJsonForm.tsx (hunk 4 above)
Comment (olaservo) [reply client/src/components/DynamicJsonForm.tsx]: I added a better clarifying comment here.  I added this debounce logic to better handle validation while typing values in JSON editor fields, since validation was getting triggered either immediately or had to be triggered on some other action like switching modes. (Edited)

File: client/src/components/DynamicJsonForm.tsx (hunk 5 above)
Comment (olaservo) [reply client/src/components/DynamicJsonForm.tsx:154]: For some fields I wanted to allow for clearing the field (i.e. setting to undefined) when its not required by the schema.
Intention is to preserve the user's ability to have an empty string input rather than forcing a default value.

Comment (olaservo) [reply client/src/components/DynamicJsonForm.tsx:154]: For some fields I wanted to allow for clearing the field (i.e. setting to undefined) when its not required by the schema.
Intention is to preserve the user's ability to have an empty string input rather than forcing a default value.

**Review by jspahrsummers — APPROVED** [review]
LGTM!

File: client/src/utils/schemaUtils.ts [hunk 6]
```diff
@@ -13,6 +13,8 @@ export function generateDefaultValue(schema: JsonSchemaType): JsonValue {
   }
 
   if (!schema.required) {
+    if (schema.type === "array") return [];
```
Comment (yGuy) [inline client/src/utils/schemaUtils.ts:15]: @olaservo  - I think this resulted in a regression. I have a schema where an option is an array that is not required. But *if*  you specify it, it should have at least one element. 

Like this:

```
  ids: z
    .array(z.number())
    .min(1)
    .optional()
    .describe("An optional array of ID#s to search for"),
```

With the change here, I don't see how you can *not* specify an array in the UI.  This worked before, AFAICT, but broke with this change.

What do you think?
